src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp

```diff
@@ -44,42 +84,107 @@ auto DrFlacImporter::doFeatures() const -> Features { return Feature::OpenData;
 bool DrFlacImporter::doIsOpened() const { return _data; }
 
 void DrFlacImporter::doOpenData(Containers::ArrayView<const char> data) {
-    drflac* handle = drflac_open_memory(data.data(), data.size());
+
+    std::unique_ptr<drflac, DrFlacDeleter> handle(drflac_open_memory(data.data(), data.size()));
     if(!handle) {
         Error() << "Audio::DrFlacImporter::openData(): failed to open and decode FLAC data";
         return;
     }
 
     uint64_t samples = handle->totalSampleCount;
-    uint32_t frequency = handle->sampleRate;
     uint8_t numChannels = handle->channels;
     uint8_t bitsPerSample = handle->bitsPerSample;
 
-    _frequency = frequency;
-
-    if(numChannels == 1 && bitsPerSample == 8)
-        _format = Buffer::Format::Mono8;
-    else if(numChannels == 1 && bitsPerSample == 16)
-        _format = Buffer::Format::Mono16;
-    else if(numChannels == 2 && bitsPerSample == 8)
-        _format = Buffer::Format::Stereo8;
-    else if(numChannels == 2 && bitsPerSample == 16)
-        _format = Buffer::Format::Stereo16;
-    else {
+    // FLAC supports any bitspersample from 4 to 64, but DrFlac always gives us 32-bit samples
+    // So we normalize bit amounts to multiples of 8, rounding up
+    UnsignedInt normalizedBytesPerSample = (bitsPerSample + 7)/8;
+
+    if(numChannels == 0 || numChannels == 3 || numChannels == 5 || numChannels > 8 ||
+       normalizedBytesPerSample == 0 || normalizedBytesPerSample > 8) {
         Error() << "Audio::DrFlacImporter::openData(): unsupported channel count"
                 << numChannels << "with" << bitsPerSample
                 << "bits per sample";
+        return;
+    }
 
-        drflac_close(handle);
+    if(samples == 0) {
+        Error() << "Audio::DrFlacImporter::openData(): no samples";
         return;
     }
 
-    Containers::Array<char> tempData(samples*sizeof(int32_t));
+    _frequency = handle->sampleRate;
+    _format = flacFormatTable[numChannels-1][normalizedBytesPerSample-1];
+    CORRADE_INTERNAL_ASSERT(_format != Buffer::Format{});
+
+    Containers::Array<char> tempData(samples*sizeof(Int));
+    drflac_read_s32(handle.get(), samples, reinterpret_cast<Int*>(tempData.begin()));
+
+    _data = convert32PCM(tempData, samples, normalizedBytesPerSample);
+
+    // 8-bit needs to become unsigned
+    if(normalizedBytesPerSample == 1) {
+        // Convert to unsigned
+        for(char& item : _data) {
+            item = item - 128;
+        }
+    }
+
+    // 24-bit needs to become float
+    else if(normalizedBytesPerSample == 3) {
+        Containers::Array<Float> floatData(samples);
+
+        for (UnsignedInt i = 0; i < samples; ++i) {
+
+            UnsignedInt s0 = _data[i*3 + 0];
+            UnsignedInt s1 = _data[i*3 + 1];
+            UnsignedInt s2 = _data[i*3 + 2];
+
+            Int sample32 = (Int)(s0 << 8) | (s1 << 16) | (s2 << 24);
+            floatData[i] = (Float)(sample32 / 2147483648.0);
```

Comment:
There's a [Math::normalize()](http://mosra.cz/blog/magnum-doc/namespaceMagnum_1_1Math.html#a767c5cede6c4e244234055ed897fec84) function for this:

```cpp
floatData[i] = Math::normalize<Float>(sample32);
```

Similarly for the other two cases below.